fix(forms): always call `setDisabledState`, not just when the control is disabled

dylhunn opened the pull request:
When a CVA is added to a form, reactive forms should always call the setDisabledState method, not just when it's disabled.

This PR provides an option in the forms module config to opt-out, since it's a breaking change. We will likely use that option inside of g3 if too many tests are broken.

Fixes #35309.

AndrewKushnir commented on goldens/public-api/forms/index.md:
```diff
@@ -662,12 +665,12 @@ export class NgForm extends ControlContainer implements Form, AfterViewInit {
 // @public
 export class NgModel extends NgControl implements OnChanges, OnDestroy {
-    constructor(parent: ControlContainer, validators: (Validator | ValidatorFn)[], asyncValidators: (AsyncValidator | AsyncValidatorFn)[], valueAccessors: ControlValueAccessor[], _changeDetectorRef?: ChangeDetectorRef | null | undefined);
+    constructor(parent: ControlContainer, validators: (Validator | ValidatorFn)[], asyncValidators: (AsyncValidator | AsyncValidatorFn)[], valueAccessors: ControlValueAccessor[], _changeDetectorRef?: ChangeDetectorRef | null | undefined, callSetDisabledState?: SetDisabledStateOption | null | undefined);
```
We'd also need to expose the `SetDisabledStateOption` symbol as a public API if we keep this approach (see another comment with a proposal to make it boolean).

dylhunn replied:
As above, I think we should keep the descriptive value names. But I have made this symbol exported.

AndrewKushnir commented on packages/forms/src/directives/shared.ts:
```diff
@@ -36,7 +52,10 @@ export function controlPath(name: string|null, parent: ControlContainer): string
+export function setUpControl(
+    control: FormControl, dir: NgControl,
+    callSetDisabledState: SetDisabledStateOption|null = setDisabledStateDefault): void {
+  if (callSetDisabledState == null) callSetDisabledState = setDisabledStateDefault;
```
```suggestion
  if (callSetDisabledState === null) callSetDisabledState = setDisabledStateDefault;
```
(unless `==` was intentional to cover the `undefined` case as well)

AndrewKushnir commented on packages/forms/src/directives/shared.ts:
```diff
@@ -46,8 +65,8 @@ export function setUpControl(control: FormControl, dir: NgControl): void {
 
-  if (control.disabled) {
-    dir.valueAccessor!.setDisabledState?.(true);
+  if (control.disabled || callSetDisabledState === 'always') {
```
I'd propose adding a quick comment here to explain the old/new behavior (for future context).

AndrewKushnir commented on packages/forms/test/value_accessor_integration_spec.ts:
```diff
@@ -1066,6 +1066,40 @@ import {dispatchEvent} from '@angular/platform-browser/testing/src/browser_util'
              expect(fixture.componentInstance.control.status).toEqual('DISABLED');
            });
 
+        describe('should support custom accessors with setDisabledState - formControlName', () => {
```
It looks like we don't use `CALL_SET_DISABLED_STATE` or `withConfig` anywhere in the test, so the current logic remains active. Could you please add a couple tests to verify that the behavior changes depending on the configuration passed using the `FormsModule.withConfig` and `ReactiveFormsModule.withConfig` calls?

dylhunn replied:
I have added an additional test to check that the `.withConfig` option can be used to turn off the fix.

AndrewKushnir commented on packages/forms/src/directives/shared.ts:
```diff
@@ -25,6 +25,23 @@ import {FormArrayName} from './reactive_directives/form_group_name';
+    'CallSetDisabledState', {providedIn: 'root', factory: () => setDisabledStateDefault});
+
+/**
+ * The type for CALL_SET_DISABLED_STATE.
+ */
```
```suggestion
 * The type for CALL_SET_DISABLED_STATE.
 * @publicApi
 */
```

AndrewKushnir replied:
Let's also add references to `withConfig` functions, so that developers can navigate to other pages that explain how/when to use the type.

atscott replied:
Does this one need to be in the public API? I don't think this appears in the public API right now.

AndrewKushnir commented on packages/forms/test/value_accessor_integration_spec.ts:
```diff
@@ -1163,6 +1197,42 @@ import {dispatchEvent} from '@angular/platform-browser/testing/src/browser_util'
+      TestBed.configureTestingModule({
+        declarations: [component, ...directives],
+        imports: [
+          FormsModule.withConfig({callSetDisabledState: 'whenDisabledForLegacyCode'}),
+          ReactiveFormsModule.withConfig({callSetDisabledState: 'whenDisabledForLegacyCode'})
```
I'd propose having 2 separate tests with 1 `withConfig` call in each, so that we verify that the DI token is added correctly (and the logic reflects that). Currently, the effect of the second call overrides the first one.

AndrewKushnir commented on packages/forms/test/value_accessor_integration_spec.ts:
```diff
@@ -1163,6 +1197,42 @@ import {dispatchEvent} from '@angular/platform-browser/testing/src/browser_util'
   });
 }
 
+{
```
```suggestion
```

AndrewKushnir reviewed (approved):
LGTM, just some minor comments below. Thank you.

atscott commented on packages/forms/src/directives/shared.ts:
```diff
@@ -25,6 +25,23 @@ import {FormArrayName} from './reactive_directives/form_group_name';
+/**
+ * The type for CALL_SET_DISABLED_STATE.
+ */
+export type SetDisabledStateOption = 'whenDisabledForLegacyCode'|'always';
```
I think this needs to be marked as publicApi and exported from the index for the docs on aio to work

atscott commented on packages/forms/src/directives/shared.ts:
```diff
@@ -25,6 +25,23 @@ import {FormArrayName} from './reactive_directives/form_group_name';
+/**
+ * The type for CALL_SET_DISABLED_STATE.
+ */
+export type SetDisabledStateOption = 'whenDisabledForLegacyCode'|'always';
```
Should we make this an enum? The `'whenDisabledForLegacyCode'` is especially long-ish

dylhunn replied:
I'd actually recommend sticking with the sum type for the sake of consistency, because that's how the other config option works. It should still be fully type-safe, since we're using this named type everywhere (not `string`).

dylhunn committed:
fix(forms): call `setDisabledState` on `ControlValueAcessor` when control is enabled

Previously, `setDisabledState` was never called when attached if the control is enabled. This PR fixes the bug, and creates a configuration option to opt-out of the fix.

Fixes #35309.

BREAKING CHANGE: setDisabledState will always be called when a `ControlValueAccessor` is attached. You can opt-out with `FormsModule.withConfig` or `ReactiveFormsModule.withConfig`.

dylhunn commented:
Global presubmit with the fix **enabled** (expected to be very breaking): [link](https://test.corp.google.com/OCL:479496210:BASE:479496227:1665121432493:b8c77b87)
Global presubmit with the fix **disabled**: [link](https://test.corp.google.com/OCL:479496364:BASE:479496492:1665121571126:c3dfb624)

jessicajaniuk reviewed (approved):
reviewed-for: public-api

AndrewKushnir commented on packages/forms/src/directives/shared.ts:
```diff
@@ -25,6 +25,27 @@ import {FormArrayName} from './reactive_directives/form_group_name';
+    'CallSetDisabledState', {providedIn: 'root', factory: () => setDisabledStateDefault});
+
+/**
+ * The type for CALL_SET_DISABLED_STATE.
```
I think we need a bit more docs here to explain what this type represents. Currently the docs refer to the `CALL_SET_DISABLED_STATE`, which is not a part of the public API.

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

dylhunn commented:
merge-assistance: This PR must be submitted with [this g3 local mod](https://critique.corp.google.com/cl/480105160) to disable the fix internally..

dylhunn commented:
State -- blocked on [a failure](https://github.com/angular/components/blob/main/src/material/legacy-chips/chip-list.spec.ts#L806) in the components repo integration test. Should be fixable though.

dylhunn commented:
The components breakage has now been fixed in https://github.com/angular/components/pull/25790

dylhunn commented:
merge-assistance: wait for CI to go green, then submit this together with [g3 local mod](https://critique.corp.google.com/cl/480151343)

devversion reviewed (approved):
LGTM for dev-infra

jessicajaniuk reviewed (approved):
reviewed-for: size-tracking

jessicajaniuk commented:
This PR was merged into the repository by commit 96b7fe93af361a1cf2ea5477970f64ba6f3d8cd5.
